File: android/app/src/main/java/com/cashu/me/Core/FocusedMintQuoteMonitor.kt
package com.cashu.me.Core

import com.cashu.me.Models.MintQuoteInfo
import com.cashu.me.Models.PaymentMethodKind
import java.util.concurrent.atomic.AtomicInteger
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive

/** The visible payment code owns the loop; cancellation releases maintenance. */
internal class FocusedMintQuoteMonitor {
    private val sessions = AtomicInteger(0)
    val isActive: Boolean get() = sessions.get() > 0

    suspend fun monitor(
        quoteId: String,
        refresh: suspend (String) -> MintQuoteInfo?,
        sleep: suspend (Long) -> Unit = { delay(it) },
    ) {
        if (!currentCoroutineContext().isActive) return
        sessions.incrementAndGet()
        try {
            while (currentCoroutineContext().isActive) {
                // Direct reconciliation bypasses passive batch/age scheduling,
                // while the reconciler still applies persisted failure backoff.
                val quote = refresh(quoteId)
                if (!currentCoroutineContext().isActive) return
                if (quote != null && quote.paymentMethod != PaymentMethodKind.Bolt12 &&
                    (quote.hasSettledPayment ||
                        (quote.paymentMethod == PaymentMethodKind.Bolt11 &&
                            quote.isExpired && quote.mintableAmount == 0L))
                ) return
                sleep(if (quote?.paymentMethod == PaymentMethodKind.Onchain) 10_000 else 2_000)
            }
        } finally {
            sessions.decrementAndGet()
        }
    }
}

File: ios/CashuWallet/Core/Services/FocusedMintQuoteMonitor.swift
import Foundation

/// The visible payment code owns this loop. Cancelling its view task releases
/// the focus, allowing ordinary wallet maintenance to resume.
@MainActor
final class FocusedMintQuoteMonitor {
    private var sessions: Set<UUID> = []
    var isActive: Bool { !sessions.isEmpty }

    func monitor(
        quoteID: String,
        refresh: (String) async -> MintQuoteInfo?,
        sleep: (Duration) async throws -> Void = { try await Task.sleep(for: $0) }
    ) async {
        guard !Task.isCancelled else { return }
        let session = UUID()
        sessions.insert(session)
        defer { sessions.remove(session) }

        while !Task.isCancelled {
            // Direct reconciliation bypasses the passive batch/age schedule,
            // but still respects persisted failure backoff in the reconciler.
            let quote = await refresh(quoteID)
            guard !Task.isCancelled else { return }
            if let quote, quote.paymentMethod != .bolt12,
               quote.hasSettledPayment ||
                (quote.paymentMethod == .bolt11 && quote.isExpired && quote.mintableAmount == 0) {
                return
            }
            do {
                try await sleep(quote?.paymentMethod == .onchain ? .seconds(10) : .seconds(2))
            } catch {
                return
            }
        }
    }
}
